File: test/test_partition.cpp
#include <stdio.h>
#include "TRTCContext.h"
#include "DVVector.h"
#include "partition.h"

int main()
{
	TRTCContext ctx;

	Functor is_even = { ctx, {},{ "x" }, "        return x % 2 == 0;\n" };

	{
		int h_value[10] = { 1, 2, 3, 4, 5, 6, 7, 8, 9, 10 };
		DVVector d_value(ctx, "int32_t", 10, h_value);
		TRTC_Partition(ctx, d_value, is_even);
		d_value.to_host(h_value);
		printf("%d %d %d %d %d ", h_value[0], h_value[1], h_value[2], h_value[3], h_value[4]);
		printf("%d %d %d %d %d\n", h_value[5], h_value[6], h_value[7], h_value[8], h_value[9]);
	}

	{
		int h_value[10] = { 0, 1, 0, 1, 0, 1, 0, 1, 0, 1 };
		DVVector d_value(ctx, "int32_t", 10, h_value);
		int h_stencil[10] = { 1, 2, 3, 4, 5, 6, 7, 8, 9, 10 };
		DVVector d_stencil(ctx, "int32_t", 10, h_stencil);
		TRTC_Partition_Stencil(ctx, d_value, d_stencil, is_even);
		d_value.to_host(h_value);
		printf("%d %d %d %d %d ", h_value[0], h_value[1], h_value[2], h_value[3], h_value[4]);
		printf("%d %d %d %d %d\n", h_value[5], h_value[6], h_value[7], h_value[8], h_value[9]);
	}

	{
		int h_value[10] = { 1, 2, 3, 4, 5, 6, 7, 8, 9, 10 };
		DVVector d_value(ctx, "int32_t", 10, h_value);
		int h_evens[10];
		DVVector d_evens(ctx, "int32_t", 10);
		int h_odds[10];
		DVVector d_odds(ctx, "int32_t", 10);
		uint32_t count = TRTC_Partition_Copy(ctx, d_value, d_evens, d_odds, is_even);
		d_evens.to_host(h_evens, 0, count);
		d_odds.to_host(h_odds, 0, 10 - count);
		for (uint32_t i = 0; i < count; i++)
			printf("%d ", h_evens[i]);
		puts("");
		for (uint32_t i = 0; i < 10-count; i++)
			printf("%d ", h_odds[i]);
		puts("");
	}

	{
		int h_value[10] = { 1, 2, 3, 4, 5, 6, 7, 8, 9, 10 };
		DVVector d_value(ctx, "int32_t", 10, h_value);
		int h_stencil[10] = { 0, 1, 0, 1, 0, 1, 0, 1, 0, 1 };
		DVVector d_stencil(ctx, "int32_t", 10, h_stencil);
		int h_evens[10];
		DVVector d_evens(ctx, "int32_t", 10);
		int h_odds[10];
		DVVector d_odds(ctx, "int32_t", 10);
		uint32_t count = TRTC_Partition_Copy_Stencil(ctx, d_value, d_stencil, d_evens, d_odds, Functor("Identity"));
		d_evens.to_host(h_evens, 0, count);
		d_odds.to_host(h_odds, 0, 10 - count);
		for (uint32_t i = 0; i < count; i++)
			printf("%d ", h_evens[i]);
		puts("");
		for (uint32_t i = 0; i < 10 - count; i++)
			printf("%d ", h_odds[i]);
		puts("");
	}

	return 0;
}

File: CSharp/test/test_partition/test_partition.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using ThrustRTCSharp;

namespace test_partition
{
    class test_partition
    {
        static void Main(string[] args)
        {
            Functor is_even = new Functor( new string[]{ "x" }, "        return x % 2 == 0;\n" );

            {
                DVVector d_value = new DVVector(new int[] { 1, 2, 3, 4, 5, 6, 7, 8, 9, 10 });
                TRTC.Partition(d_value, is_even);
                print_array((int[])d_value.to_host());
            }

            {
                DVVector d_value = new DVVector(new int[] { 0, 1, 0, 1, 0, 1, 0, 1, 0, 1 });
                DVVector d_stencil = new DVVector(new int[] { 1, 2, 3, 4, 5, 6, 7, 8, 9, 10 });
                TRTC.Partition_Stencil(d_value, d_stencil, is_even);
                print_array((int[])d_value.to_host());
            }

            {
                DVVector d_value = new DVVector(new int[] { 1, 2, 3, 4, 5, 6, 7, 8, 9, 10 });
                DVVector d_evens = new DVVector("int32_t", 10);
                DVVector d_odds = new DVVector("int32_t", 10);
                long count = TRTC.Partition_Copy(d_value, d_evens, d_odds, is_even);
                print_array((int[])d_evens.to_host(0, count));
                print_array((int[])d_odds.to_host(0, 10 - count));
            }

            {
                DVVector d_value = new DVVector(new int[] { 1, 2, 3, 4, 5, 6, 7, 8, 9, 10 });
                DVVector d_stencil = new DVVector(new int[] { 0, 1, 0, 1, 0, 1, 0, 1, 0, 1 });
                DVVector d_evens = new DVVector("int32_t", 10);
                DVVector d_odds = new DVVector("int32_t", 10);
                long count = TRTC.Partition_Copy_Stencil(d_value, d_stencil, d_evens, d_odds, new Functor("Identity"));
                print_array((int[])d_evens.to_host(0, count));
                print_array((int[])d_odds.to_host(0, 10 - count));
            }

            {
                DVVector d_value = new DVVector(new int[] { 2, 4, 6, 8, 10, 1, 3, 5, 7, 9 });
                Console.WriteLine(TRTC.Partition_Point(d_value, is_even));
            }
        }

        static void print_array<T>(T[] arr)
        {
            foreach (var item in arr)
            {
                Console.Write(item.ToString());
                Console.Write(" ");
            }
            Console.WriteLine("");
        }
    }
}
